step-grid-parent/step-grid-server/src/main/java/step/grid/GridImpl.java

```diff
@@ -242,11 +291,35 @@ private void startServer() throws Exception {
 	}
 	
 	protected void handleRegistrationMessage(RegistrationMessage message) {
+		if (!acceptRegistrationMessages) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("Currently not accepting registration messages, ignoring.");
+			}
+			return;
+		}
+
 		AgentRef agentRef = message.getAgentRef();
-		agentRefs.putOrTouch(agentRef.getAgentId(), agentRef);
-		for (Token token : message.getTokens()) {
-			tokenPool.offerToken(new TokenWrapper(token, agentRef));			
-		}	
+		boolean allowed = agentRegistrationCallbacks.stream().allMatch(cb -> cb.beforeRegistering(agentRef));
+		if (logger.isDebugEnabled()) {
+			if (!allowed) {
+				logger.debug("One or more callbacks vetoed agent registration, ignoring agent: {}", agentRef);
+			} else {
+				logger.debug("Allowing agent: {}", agentRef);
+			}
+		}
+		if (allowed) {
+			agentRefs.putOrTouch(agentRef.getAgentId(), agentRef);
+			for (Token token : message.getTokens()) {
+				tokenPool.offerToken(new TokenWrapper(token, agentRef));
+			}
+		} else {
+			if (agentRefs.remove(agentRef.getAgentId()) != null) {

```

Comment:
@cl-exense The agent wouldn't be informed that his registration request has been refused. Should we throw an exception in this case to return an error to the rest service?

Reply:
I just tried that out. The effect is that on the *server* side, the log is littered with stacktraces every second, whereas on the agent side, nothing is logged at all (at least with the default settings, not sure what happens when cranking up debugging) :-D